### PR TITLE
Modularize vtd 3

### PR DESCRIPTION
**Pull request by rarindam**

### CODE REVIEW

**Comment by HuiHuangShi** [comment]
2361<br>1 new potential violations to the coding guideline detected. Please help check. Thanks!  <br><br>1.filename:/hypervisor/arch/x86/vtd.c&nbsp;&nbsp;function:None&nbsp;&nbsp;offset:150:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:<#> Identifier reuse: persistent var vs var. : vm0_domain<br><br>

**Comment by rarindam** [comment]
It builds successfully for all targets, but fails here in mrb:
[2019-01-17 08:22:02.812] _launch_uos_one_time (<device.mrb.MRB object at 0x7f188c96f710>, 'LaaG', 'serial', '/usr/share/acrn/samples/apl-mrb/launch_uos.sh', 300) {'ssh_laag_flag': False}
[ ERROR ] write: sh /usr/share/acrn/samples/apl-mrb/launch_uos.sh -V 1
 and expect: login: failed with timeout: 300
sh /usr/share/acrn/samples/apl-mrb/launch_uos.sh -V 1

I do not see this issue, when running the test build myself.

**Comment by wenlingz** [comment]
@rarindam please check your code, fail to launch UOS, please refer the log:
********************************************************************************
root@clr-8bac1fa012b8482bb238d1e6ada085bb ~ # /usr/share/acrn/samples/apl-mrb/launch_uos.sh -V 2
cpu1 online=1
cpu2 online=1
cpu3 online=1
Launch android UOS
acrn-br0 bridge aleady exists, adding new tap device to it...
       -G: GVT args: low_gm_size, high_gm_size, fence_sz
passed gvt-g optargs low_gm 64, high_gm 448, fence 8
interrupt storm monitor params: 10000, 10, 1, 100
SW_LOAD: get vsbl path /usr/share/acrn/bios/VSBL.bin
SW_LOAD: get bootargs maxcpus=3 nohpet tsc=reliable intel_iommu=off    androidboot.serialno=R1J56La8b3126d    i915.enable_rc6=1 i915.enable0
VHM api version 1.0
open hugetlbfs file /run/hugepage/acrn/huge_lv1/D279543825D611E8864ECB7A18B34643
open hugetlbfs file /run/hugepage/acrn/huge_lv2/D279543825D611E8864ECB7A18B34643
level 0 free/need pages:0/0 page size:0x200000
level 1 free/need pages:0/2 page size:0x40000000
to reserve more free pages:
to reserve pages (+orig 0): echo 2 > /sys/kernel/mm/hugepages/hugepages-1048576kB/nr_hugepages
to reserve pages (+orig 0): echo 512 > /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages
now enough free pages are reserved!

try to setup hugepage with:
        level 0 - lowmem 0x0, biosmem 0x0, highmem 0x0
        level 1 - lowmem 0x80000000, biosmem 0x0, highmem 0x0
total_size 0xc0000000

mmap ptr 0x0x7f77fdec1000 -> baseaddr 0x0x7f7800000000
mmap 0x40000000@0x7f7800000000
touch 1 pages with pagesz 0x40000000
mmap 0x40000000@0x7f7840000000
touch 512 pages with pagesz 0x200000

really setup hugepage with:
        level 0 - lowmem 0x40000000, biosmem 0x0, highmem 0x0
        level 1 - lowmem 0x40000000, biosmem 0x0, highmem 0x0
start monitor interrupt data...
Listening 27...
polling 27...
[101440552us][cpu=0][sev=3][seq=46]:iommu: add_iommu_device: context entry@0x74c65a90 (Lower:74c44001)
[101450888us][cpu=0][sev=3][seq=47]:iommu: already present for 0:15.1
[101472835us][cpu=0][sev=3][seq=48]:iommu: remove_iommu_device: domain id mismatch
acrn-dm: PCI device at 0/15/1 is not using the pt(4) driver
mngr_client_new 113
Stop listening 27...
Stop polling 27...
Unable to init vdev (9)

**Comment by wenlingz** [comment]
@rarindam please also add the Tracked-On info and acked-by info, thanks

**Comment by binbinwu1** [comment]
I am OK with the change, but didn't find the reason why the test failed. According to the log, when start UOS, a passthru device need to be removed from domain0 and added to domainU, but according to the vtd table, the device is not in domain 0 already.
@wenlingz Is the log you posted here the log from UOS auto boot or you tested it manually?

**Comment by binbinwu1** [comment]
> @binbinwu1 the log which we test it manually

Is there an autoboot UOS launch before your manual test?

**Comment by wenlingz** [comment]
@rarindam please rebase your code and fix your code, which fail to launch uos,thanks

File: doc/developer-guides/hld/hv-vt-d.rst [hunk 1]
```diff
@@ -306,9 +306,6 @@ deinitialization:
 .. doxygenfunction:: init_iommu
    :project: Project ACRN
```
Comment (dbkinder) [inline doc/developer-guides/hld/hv-vt-d.rst]: I notice this function (init_iommu_vm0_domain) was moved to another file, but still exists.  Is it correct to remove it from the documentation?

**Comment by rarindam** [comment]
As more patches were merged, this patch became out dated. There has been some changes in the renaming if vm0 functions to sos.
Need to resbmit after pulling in all the changes.
Hence closing this one.

**Pull request closed without merging**